client/core/bot.go

```diff
@@ -0,0 +1,1734 @@
+
+	var lastCheck time.Time
+	m.mm.cache.RLock()
+	if p := m.mm.cache.prices[m.market.Name]; p != nil {
+		lastCheck = p.stamp
+	}
+	m.mm.cache.RUnlock()
+
+	// If we don't already have a price, sync now.
+	if time.Since(lastCheck) > time.Minute {
+		if _, err := m.syncOraclePrice(ctx); err != nil {
+			m.log.Errorf("failed to establish a market-averaged price: %v", err)
+		}
+	}
+
+	ticker := time.NewTicker(oracleRecheckInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+		case <-ctx.Done():
+			return
+		}
+		if _, err := m.syncOraclePrice(ctx); err != nil {
+			m.log.Errorf("failed to get market-averaged price: %v", err)
+		}
+	}
+}
+
+// basisPrice is the basis price for the makerBot's market.
+func (m *makerBot) basisPrice() uint64 {
+	pgm := m.program()
+	midGap, err := m.book.MidGap()
+	if err != nil && !errors.Is(err, orderbook.ErrEmptyOrderbook) {
+		m.log.Errorf("error calculating mid-gap: %w", err)
+		return 0
+	}
+	if p := basisPrice(pgm.Host, m.market, pgm.OracleBias, pgm.oracleWeighting(), midGap, m, m.log); p > 0 {
+		return p
+	}
+	return m.market.ConventionalRateToMsg(pgm.EmptyMarketRate)
+}
+
+type basisPricer interface {
+	cachedOraclePrice(mktName string) *stampedPrice
+	fiatConversions() map[uint32]float64
+}
+
+// basisPrice is the oracle-weighted price on which to base the target price and
+// spread. The oracle weighting and bias is applied, but no other adjustments.
+func basisPrice(host string, mkt *Market, oracleBias, oracleWeighting float64, midGap uint64, bp basisPricer, log dex.Logger) uint64 {
+	basisPrice := float64(midGap) // float64 message-rate units
+
+	log.Tracef("basisPrice: mid-gap price = %d", midGap)
+
+	// Our sync loop should be running. Only check the cache.
+	p := bp.cachedOraclePrice(mkt.Name)
+
+	if p != nil && p.price != 0 {
+		log.Tracef("basisPrice: raw oracle price = %.8f", p.price)
+
+		msgOracleRate := float64(mkt.ConventionalRateToMsg(p.price))
+
+		// Apply the oracle mismatch filter.
+		if basisPrice > 0 {
+			low, high := msgOracleRate*(1-maxOracleMismatch), msgOracleRate*(1+maxOracleMismatch)
+			if basisPrice < low {
+				log.Debug("local mid-gap is below safe range. Using effective mid-gap of %d%% below the oracle rate.", maxOracleMismatch*100)
+				basisPrice = low
+			} else if basisPrice > high {
+				log.Debug("local mid-gap is above safe range. Using effective mid-gap of %d%% above the oracle rate.", maxOracleMismatch*100)
+				basisPrice = high
+			}
+		}
+
+		if oracleBias != 0 {
+			msgOracleRate *= 1 + oracleBias
+
+			log.Tracef("basisPrice: biased oracle price = %.0f", msgOracleRate)
+		}
+
+		if basisPrice == 0 { // no mid-gap available. Just use the oracle value straight.
+			basisPrice = msgOracleRate
+			log.Tracef("basisPrice: using basis price %.0f from oracle because no mid-gap was found in order book", basisPrice)
+		} else if oracleWeighting > 0 {
+			basisPrice = msgOracleRate*oracleWeighting + basisPrice*(1-oracleWeighting)
```

Comment:
For a low-depth book, it would be super easy to drain a bot at a favorable price by manipulating the mid-gap rate.

Reply:
So the oracle weight makes it harder to do this type of manipulation, but if someone is able to do a major manipulation, then even a 50% weight wouldn't save the victim bots. I think it would make sense to have the oracle weight combined with another setting that says if the difference between the mid-gap rate and the oracle rate is greater than some percentage, then don't place any orders.

Reply:
> I think I've captured at least part of what you're asking with maxOracleMismatch

Ah yeah that's right. 

Another feature could be for the bot to turn off if it keeps losing money. I'd definitely have more peace of mind to leave it on after going to sleep like that.